### PR TITLE
FormRecognizer Receipt Update

### PR DESCRIPTION
**Pull request by ramparab**
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [ ] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [ ] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json [hunk 1]
```diff
@@ -87,21 +127,51 @@
...
             }
           },
-          "default": {
-            "description": "Error response.",
-            "schema": {
-              "$ref": "FormRecognizerReceiptOcr.json#/definitions/ComputerVisionError"
+          "404": {
```
Comment (yangyuan) [inline specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: I was told we should not define errors in swagger.
Otherwise the generated SDK will not throw exceptions for those errors.

Need ADX team to confirm if it still holds.

**Review by yangyuan — CHANGES_REQUESTED** [review]

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json (hunk 1 above)
Comment (bojunehsu) [reply specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: We do see errors during model validation in the PR due to the error codes.  In particular, it is expecting that we have examples for each response code.  I am okay with removing the error response codes and replacing them with the previous "default".

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json [hunk 2]
```diff
@@ -87,21 +127,51 @@
...
...gBox</td><th>[Number]</th><td>The bounding box of a recognized line or word, depending on the parent object. It is represented by an array of eight numbers corresponding to four points, clockwise from the top-left corner. For image, the (x, y) coordinates are measured in pixels from the top-left corner of the image. For PDF, the (x, y) coordinates are measured in inches from the top-left corner of the each page.</td></tr><tr><td>confidence</td><th>String</th><td>This flag appears when the recognition of a word is of low confidence.</td></tr><tr><td>width</td><th>Number</th><td>The width of the image/PDF in pixels/inches, respectively.</td></tr><tr><td>height</td><th>Number</th><td>The height of the image/PDF in pixels/inches, respectively.</td></tr><tr><td>clockwiseOrientation</td><th>Number</th><td>The orientation of the image in clockwise direction, measured in degrees between [0, 360).</td></tr><tr><td>page</td><th>Integer</th><td>The 1-based page number of the recognition result.</td></tr><tr><td>unit</td><th>String</th><td>The unit used by the width, height and boundingBox properties. For images, the unit is \"pixel\". For PDF, the unit is \"inch\".</td></tr></tbody></table>",
```
Comment (yungezz) [inline specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: swagger will be used to generate docs at docs.ms and SDKs.  xml inside json will be terrible for downstream docs and SDKs, might cause some issue there. Could you pls put sample into sample json file instead of swagger json file?

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json [hunk 3]
```diff
@@ -47,37 +48,76 @@
...
+                "type": "string",
+                "description": "URL to query and retrieve the analysis result. This URL expires in 48 hours."
               }
             }
           },
-          "default": {
-            "description": "Error response.",
-            "schema": {
-              "$ref": "FormRecognizerReceiptOcr.json#/definitions/ComputerVisionError"
+          "400": {
```
Comment (yungezz) [inline specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: pls remove error code from swagger file.

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json [hunk 4]
```diff
@@ -253,6 +483,51 @@
...
+    "Endpoint": {
+      "name": "Endpoint",
+      "description": "Supported Cognitive Services endpoints (protocol and hostname, for example: https://westus2.api.cognitive.microsoft.com).",
+      "x-ms-parameter-location": "client",
+      "required": true,
+      "type": "string",
+      "in": "path",
+      "x-ms-skip-url-encoding": true
+    },
+    "ImageUrl": {
```
Comment (yungezz) [inline specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: type

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceiptOcr.json [hunk 5]
```diff
@@ -1,222 +0,0 @@
-{
```
Comment (yungezz) [inline specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceiptOcr.json]: the file is removed intentionally? although it's a preview version, to double confirm

Comment (bojunehsu) [reply specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceiptOcr.json]: Yes.  We previously used external references.  We now merged the content into FormRecognizerReceipt.json so there are no external dependencies.

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json (hunk 4 above)
Comment (bojunehsu) [reply specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: Do we really want to specify type here when we reference the schema?

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json (hunk 2 above)
Comment (bojunehsu) [reply specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: Got it.  This makes it nearly impossible to share Swagger with APIM documentation.  We will place the sample in the external x-ms-examples file.

**Review by yangyuan — APPROVED** [review]

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json [hunk 6]
```diff
@@ -32,11 +32,12 @@
   "paths": {
     "/prebuilt/receipt/asyncBatchAnalyze": {
       "post": {
-        "description": "Batch Read Receipt operation. The response contains a field called 'Operation-Location', which contains the URL that you must use for your 'Get Read Receipt Result' operation.",
-        "operationId": "BatchReadReceipt",
+        "description": "Extract field text and semantic values from a given receipt document.",
+        "operationId": "AnalyzeReceipt",
```
Comment (johanste) [inline specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: Was this rename intentional? It will be a breaking change for autorest generated client libraries...

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json [hunk 7]
```diff
@@ -47,37 +48,32 @@
...
             "headers": {
               "Operation-Location": {
-                "description": "URL to query for status of the operation. The URL will expire in 48 hours. ",
-                "type": "string"
+                "type": "string",
+                "description": "URL to query and retrieve the analysis result. This URL expires in 48 hours."
               }
             }
-          },
-          "default": {
```
Comment (johanste) [inline specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: Why are you removing the schema information for default/non-success status codes?

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json (hunk 6 above)
Comment (bojunehsu) [reply specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: Yes it is intentional.  We have not released the client libraries because the name was inconsistent before.

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json (hunk 7 above)
Comment (bojunehsu) [reply specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: Previous reviewers indicated that explicitly listing individual error status codes (4xx, 5xx) may cause issues in the SDK generation.  While removing the individual status codes, I forgot to change it back to default.  Is it critical that we specify the schema for default response code?

One question: If we explicitly setting the schema for default status code, it means that we cannot have different schemas across error codes.  Though that is currently not an issue in our current implementation, it feels odd.  What is the guideline here?

Comment (johanste) [reply specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: The general guideline is to (assuming that you are using the same schema for non-success status code - which the vast majority of services do), provide said schema in the "default" section.

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json [hunk 8]
```diff
@@ -125,47 +116,161 @@
...
+        "lines"
+      ],
+      "properties": {
+        "page": {
+          "description": "The 1-based page number of the recognition result.",
+          "type": "integer"
+        },
+        "clockwiseOrientation": {
+          "description": "The orientation of the image in clockwise direction, measured in degrees between [0, 360).",
+          "type": "number"
```
Comment (johanste) [inline specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: You can specify `minimum`, `maximum` and `exclusiveMaximum` values to indicate the range of values.

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json [hunk 9]
```diff
@@ -125,47 +116,161 @@
...
+        "unit": {
+          "description": "The unit used by the width, height and boundingBox properties. For images, the unit is \"pixel\". For PDF, the unit is \"inch\".",
+          "type": "string",
+          "enum": [
+            "pixel",
+            "inch"
+          ],
+          "x-ms-enum": {
+            "name": "TextRecognitionResultDimensionUnit",
+            "modelAsString": false
```
Comment (johanste) [inline specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: Explicitly marking something as `"modelAsString": false` is not required (it is the default). But please make sure that this is intentional from a versioning perspective.

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json [hunk 10]
```diff
@@ -125,47 +116,161 @@
...
+      "x-ms-enum": {
+        "name": "OperationStatusCodes",
+        "modelAsString": false
+      },
+      "x-nullable": false
+    },
+    "TextRecognitionResult": {
+      "description": "Text recognition result from a page in the input document.",
+      "type": "object",
+      "required": [
```
Comment (johanste) [inline specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: Is this the only required value?

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json [hunk 11]
```diff
@@ -125,47 +116,161 @@
...
+            "name": "TextRecognitionResultConfidenceClass",
+            "modelAsString": false
+          },
+          "x-nullable": true
+        }
+      }
+    },
+    "BoundingBox": {
+      "description": "Quadrangle bounding box, with coordinates specified relative to the original image. The eight numbers represent the four points, clockwise from the top-left corner. For image, the (x, y) coordinates are measured in pixels. For PDF, the (x, y) coordinates are measured in inches.",
+      "type": "array",
```
Comment (johanste) [inline specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: Have you considered having four points instead of an array with eight numbers?

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json [hunk 12]
```diff
@@ -125,47 +116,161 @@
...
+      "items": {
```
Comment (johanste) [inline specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: Please specify min/maxLength...

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json [hunk 13]
```diff
@@ -253,6 +359,51 @@
...
+      }
+    },
+    "ImageStream": {
+      "name": "Image",
+      "in": "body",
+      "required": true,
+      "x-ms-parameter-location": "method",
+      "description": "An image stream.",
+      "schema": {
+        "type": "object",
```
Comment (johanste) [inline specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: `"type": "file"`?

**Review by johanste — CHANGES_REQUESTED** [review]
A couple of general comments:

* Please verify that you really intend to specify `"modelAsString": false` 
* Please double-check that you have specified all parameters that are `required` as, well, `required`.
* Numerical values that have a contractual/natural range should have min/max values specified (e.g. it is unlikely that you will support ranges other than 0-360 degrees for rotation angles). This helps both with documentation and client side validation.

Also, please add the "default" sections back.

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json (hunk 11 above)
Comment (bojunehsu) [reply specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: The current design is selected to be consistent with the GAed OCR service.

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json (hunk 12 above)
Comment (bojunehsu) [reply specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: Interpreting this as min/maxItems.

File: specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json (hunk 13 above)
Comment (bojunehsu) [reply specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: We are using the same backend service as OCR, whose Swagger is currently defined using type object and application/octet-stream instead of the recommendations in https://swagger.io/docs/specification/2-0/file-upload/ with type file and multipart/form-data.  I will test to see if the recommendations are compatible with the current backend service and update if it works.

Comment (johanste) [reply specification/cognitiveservices/data-plane/FormRecognizer/preview/v1.0/FormRecognizerReceipt.json]: What is the intent behind the `"format": "file"` addition? `"file"` is a type.  Did you mean to specify `"type": "binary"`?

**Comment by yungezz** [comment]
HI @ramparab , are you still on this PR? could you pls address comments?

**Comment by yungezz** [comment]
HI @ramparab , are you still on this PR? could you pls address comments? Is it ok for close the PR if you're not ready for changing it?

**Comment by yungezz** [comment]
HI @ramparab , are you still on this PR? could you pls address comments? Is it ok for close the PR if you're not ready for changing it?

**Comment by yungezz** [comment]
close after email communication with PR authors.
